src/main/java/org/visola/cs/datastructure/CircularQueue.java

```diff
@@ -0,0 +1,87 @@
+package org.visola.cs.datastructure;
+
+public class CircularQueue<T> implements Queue<T> {
+
+  private final Object[] items;
+
+  private int head = -1;
```

Comment:
I know `head` and `tail` are commonly used to denote the start and end positions for data structures, but since the API is called `front` and `rear` maybe it would make sense for these instance variables to reflect that naming convention?

Also, these represent the index positions for the respective `front` and `rear` elements. Maybe it's superfluous but you could consider naming these values `frontIndex` and `rearIndex` to make it abundantly clear what they represent.

Reply:
I thought about how to name these things, but the interface was declared in the problem. And the names for these were stated in the problem (they are not accessible outside the class and not evaluated as part of the problem).
But since this is a common/known problem, I prefer to stick with these names since they are not ambiguous (type is `int` and type for actual value of `head` is `T`).
I'd rather call them `headIndex` and `tailIndex` to make it "abundantly clear". :-P

Reply:
Ok, I was looking at the JavaScript example which has less constraints 😉